checker/src/main/java/org/checkerframework/checker/signedness/SignednessVisitor.java

```diff
@@ -158,6 +166,63 @@ public Void visitBinary(BinaryTree node, Void p) {
     return super.visitBinary(node, p);
   }
 
+  // Ensure that method annotations are not written on methods they don't apply to.
+  // Copied from InterningVisitor
+  @Override
+  public Void visitMethod(MethodTree node, Void p) {
+    ExecutableElement methElt = TreeUtils.elementFromDeclaration(node);
+    boolean hasEqualsMethodAnno =
+        atypeFactory.getDeclAnnotation(methElt, EqualsMethod.class) != null;
+    int params = methElt.getParameters().size();
+    if (hasEqualsMethodAnno && !(params == 1 || params == 2)) {
+      checker.reportError(
+          node, "invalid.method.annotation", "@EqualsMethod", "1 or 2", methElt, params);
+    }
+
+    return super.visitMethod(node, p);
+  }
+
+  /**
+   * If lint option "dotequals" is specified, warn if the .equals method is used where reference
+   * equality is safe.
+   */
+  @Override
+  public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+    ExecutableElement methElt = TreeUtils.elementFromUse(node);
+    boolean hasEqualsMethodAnno =
+        atypeFactory.getDeclAnnotation(methElt, EqualsMethod.class) != null;
+    if (hasEqualsMethodAnno || InterningVisitor.isInvocationOfEquals(node)) {
+      int params = methElt.getParameters().size();
+      if (!(params == 1 || params == 2)) {
+        checker.reportError(
+            node, "invalid.method.annotation", "@EqualsMethod", "1 or 2", methElt, params);
+      } else {
+        AnnotatedTypeMirror leftOpType;
+        AnnotatedTypeMirror rightOpType;
+        if (params == 1) {
+          leftOpType = atypeFactory.getReceiverType(node);
+          rightOpType = atypeFactory.getAnnotatedType(node.getArguments().get(0));
+        } else if (params == 2) {
+          leftOpType = atypeFactory.getAnnotatedType(node.getArguments().get(0));
+          rightOpType = atypeFactory.getAnnotatedType(node.getArguments().get(1));
+        } else {
+          throw new BugInCF("Checked that params is 1 or 2");
+        }
+        if (!atypeFactory.maybeIntegral(leftOpType) || !atypeFactory.maybeIntegral(rightOpType)) {
+          // nothing to do
+        } else if (leftOpType.hasAnnotation(Unsigned.class)
+            && rightOpType.hasAnnotation(Signed.class)) {
```

Comment:
If the rightOp is `@SignednessGlb` or any other subtype of `@Signed` then an error won't be issued.

Reply:
Every subtype of `@Signed` is also a subtype of `@Unsigned`, so I think this is OK.